Static Graph RestoreEx: NuGet.Build.Tasks.Console.exe crash with wildcards and UNC paths

I have a situation where normal restore works fine, but switching to static graph restore crashes MSBuild with:

`The UNC path should be of the form \\server\share.`

```
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): error : The UNC path should be of the form \\server\share. [C:\Ide\Ide.sln]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): error :    at System.IO.LongPathHelper.Normalize(String path, UInt32 maxPathLength, Boolean checkInvalidCharacters, Boolean expandShortPaths) [C:\Ide\Ide.sln]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): error :    at System.IO.Path.NormalizePath(String path, Boolean fullCheck, Int32 maxPathLength, Boolean expandShortPaths) [C:\Ide\Ide.sln]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): error :    at System.IO.Path.GetFullPathInternal(String path) [C:\Ide\Ide.sln]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): error :    at Microsoft.Build.Shared.FileUtilities.GetFullPath(String path) [C:\Ide\Ide.sln]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): error :    at Microsoft.Build.Shared.FileUtilities.NormalizePath(String path) [C:\Ide\Ide.sln]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): error :    at Microsoft.Build.Shared.FileUtilities.GetFullPath(String fileSpec, String currentDirectory) [C:\Ide\Ide.sln]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): error :    at Microsoft.Build.Evaluation.LazyItemEvaluator`4.LazyItemList.ProcessNonWildCardItemUpdates(Dictionary`2 itemsWithNoWildcards, Builder items) [C:\Ide\Ide.sln]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): error :    at Microsoft.Build.Evaluation.LazyItemEvaluator`4.LazyItemList.ComputeItems(LazyItemList lazyItemList, ImmutableHashSet`1 globsToIgnore) [C:\Ide\Ide.sln]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): error :    at Microsoft.Build.Evaluation.LazyItemEvaluator`4.LazyItemList.GetItemData(ImmutableHashSet`1 globsToIgnore) [C:\Ide\Ide.sln]
```

The specific process that crashes is `NuGet.Build.Tasks.Console.exe` with the following stack:

```
	mscorlib	LongPathHelper.Normalize Line 167
	mscorlib	Path.NewNormalizePath Line 503
	mscorlib	Path.NormalizePath Line 427
	mscorlib	Path.NormalizePath Line 391
	mscorlib	Path.NormalizePath Line 376
	mscorlib	Path.GetFullPathInternal Line 366
	mscorlib	Path.GetFullPath Line 330
	Microsoft.Build	FileUtilities.GetFullPath Line 500
	Microsoft.Build	FileUtilities.NormalizePath Line 465
	Microsoft.Build	FileUtilities.GetFullPath Line 762
	Microsoft.Build	LazyItemEvaluator`4.LazyItemList.ProcessNonWildCardItemUpdates Line 447
	Microsoft.Build	LazyItemEvaluator`4.LazyItemList.ComputeItems Line 418
	Microsoft.Build	LazyItemEvaluator`4.LazyItemList.GetItemData Line 306
	Microsoft.Build	LazyItemEvaluator`4.GetAllItemsDeferred Line 506
	System.Core	Enumerable.SelectManyIterator
	System.Core	Buffer`1..ctor
	System.Core	OrderedEnumerable`1.GetEnumerator
	Microsoft.Build	Evaluator`4.Evaluate Line 686
	Microsoft.Build	Evaluator`4.Evaluate Line 341
	Microsoft.Build	ProjectInstance.Initialize Line 2804
	Microsoft.Build	ProjectInstance..ctor Line 285
	Microsoft.Build	ProjectInstance.FromFile Line 757
	NuGet.Build.Tasks.Console	MSBuildStaticGraphRestore.LoadProjects
	Microsoft.Build	GraphBuilder.ParseProject Line 536
	Microsoft.Build	GraphBuilder.SubmitProjectForParsing Line 576
	mscorlib	Lazy`1.CreateValue Line 437
	mscorlib	Lazy`1.LazyInitValue Line 388
	mscorlib	Lazy`1.get_Value Line 339
	Microsoft.Build	ParallelWorkSet`2.ExecuteWorkItem Line 194
	Microsoft.Build	ParallelWorkSet`2.<CreateProcessorItemTask>b__17_0 Line 169
	mscorlib	AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext Line 1090
	mscorlib	ExecutionContext.RunInternal Line 981
	mscorlib	ExecutionContext.Run Line 928
	mscorlib	AsyncMethodBuilderCore.MoveNextRunner.Run Line 1071
	mscorlib	AwaitTaskContinuation.RunOrScheduleAction Line 811
	mscorlib	Task.FinishContinuations Line 3617
	mscorlib	Task.FinishStageThree Line 2363
	mscorlib	Task`1.TrySetResult Line 490
	mscorlib	SemaphoreSlim.TaskNode.ExecuteWorkItem Line 95
	mscorlib	ThreadPoolWorkQueue.Dispatch Line 820
	mscorlib	_ThreadPoolWaitCallback.PerformWaitCallback Line 1161
```

The actual problem is this call to `Path.GetFullPath`:
https://github.com/dotnet/msbuild/blob/c36a54ed3308d1516ffe1a86b9086c42e4ca996f/src/Build/Evaluation/LazyItemEvaluator.cs#L447

It is being passed the value of `\\csc.*`, which results in GetFullPath throwing an ArgumentException: `The UNC path should be of the form \\server\share.`

Unfortunately I've failed to isolate a standalone repro for this exact scenario. I do have an internal repo at commit `bd36c7032f0950eeca634d7aa34fcdf441e89ed9` that's easy to reproduce this on. Just contact me on Teams and we can take a look together.

However I did build a repro that hits a very similar case, with slightly different behavior. It no longer crashes, but fails the build with this error:

```
"C:\temp\staticGraphFail\staticGraphFail.sln" (Restore target) (1) ->
(Restore target) ->
  C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.DefaultItems.targets(101,13): error MSB4248: Cannot expand metadata in expression "$([MSBuild]::
ValueOrDefault('%(FullPath)', '').StartsWith($([MSBuild]::EnsureTrailingSlash($(MSBuildProjectDirectory)))))". The item metadata "%(FullPath)" cannot be applied to the path "\\csc.*". The UNC path
 should be of the form \\server\share.  C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.DefaultItems.targets [C:\temp\staticGraphFail\staticGraphF
ail.sln]
```

To repro, create this fail.csproj file:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <RestoreUseStaticGraphEvaluation>true</RestoreUseStaticGraphEvaluation>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Newtonsoft.Json" Version="13.0.3" ExcludeAssets="all" GeneratePathProperty="true"/>
  </ItemGroup>

  <PropertyGroup>
    <A>$(PkgNewtonsoft_Json)\</A>
  </PropertyGroup>

  <ItemGroup>
    <None Include="$(A)\csc.*" />
  </ItemGroup>

</Project>
```

and build with `msbuild /t:Restore fail.csproj`.

In this repro, the crash happens here:
https://github.com/dotnet/msbuild/blob/c36a54ed3308d1516ffe1a86b9086c42e4ca996f/src/Build/Evaluation/LazyItemEvaluator.cs#L429

There's a good try/catch around this callstack, so it doesn't bring down the process, because this catch block catches it:
https://github.com/dotnet/msbuild/blob/c36a54ed3308d1516ffe1a86b9086c42e4ca996f/src/Shared/Modifiers.cs#L602

There's another catch block here:
https://github.com/dotnet/msbuild/blob/c36a54ed3308d1516ffe1a86b9086c42e4ca996f/src/Build/Evaluation/Expander.cs#L982

However in my actual case, there isn't such a good catch block around this codepath:
https://github.com/dotnet/msbuild/blob/c36a54ed3308d1516ffe1a86b9086c42e4ca996f/src/Build/Evaluation/LazyItemEvaluator.cs#L418

I wasn't able to recreate a standalone repro which takes that path (line 418) instead of the regular path (line 429). Ping me on Teams for a real-life repro of the actual issue.

I suspect the problem is that wildcards are not expected to be in the item spec `\\csc.*` but they are, and the item spec starts with `\\` because the $(Pkg*) property evaluates to an empty string at the time of restore eval, and then `Path.GetFullPath()` crashes on it, but there's no good catch block to catch it around line 418, so it is more fatal.

> Gets past the crash but still fails with
> 
> ```
> C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.DefaultItems.targets(101,13): error MSB4248: Cannot expand metadata in expression "$([MSBuild]::ValueOrDefault('%(FullPath)', '').StartsWith($([MSBuild]::EnsureTrailingSlash($(MSBuildProjectDirectory)))))". The item metadata "%(FullPath)" cannot be applied to the path "\\csc.*". The UNC path should be of the form \\server\share.  C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.DefaultItems.targets
> ```

This can be worked around with `SetLinkMetadataAutomatically=false`

```sh-session
❯ artifacts\bin\bootstrap\net472\MSBuild\Current\Bin\amd64\MSBuild.exe S:\repro\dotnet\msbuild\issues\9405\fail.csproj -t:restore -p:SetLinkMetadataAutomatically=false
Restore complete (1.6s)

Build succeeded in 2.1s
```

--- a/src/Build.UnitTests/Evaluation/ItemEvaluation_Tests.cs
+++ b/src/Build.UnitTests/Evaluation/ItemEvaluation_Tests.cs
@@ -9,6 +9,8 @@ using System.Text;
 using Microsoft.Build.Evaluation;
 using Microsoft.Build.Internal;
 using Microsoft.Build.Shared;
+
+using Shouldly;
 using Xunit;
 
 #nullable disable
@@ -622,6 +624,41 @@ namespace Microsoft.Build.UnitTests.Evaluation
                 // recursive dir does not work with lazy wildcards
                 Assert.Equal(string.Empty, project.GetConcatenatedItemsOfType("RecursiveDir"));
             }
+        }
+
+        [Fact]
+        public void DoesNotCrashWhenUnEvaluatedWildCardLooksLikeUNC()
+        {
+            var content = """
+                <Project>
+
+                  <PropertyGroup>
+                    <A>$(B)\</A>
+                  </PropertyGroup>
+
+                  <ItemGroup>
+                    <None Include="$(A)\csc.*" />
+                    <None Update="2.txt">
+                      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+                    </None>
+                  </ItemGroup>
+
+                  <Target Name="a" />
+
+                </Project>
+                """.Cleanup();
+
+            using var env = TestEnvironment.Create();
+
+            var projectFiles = env.CreateTestProjectWithFiles(content);
+
+            env.SetEnvironmentVariable("MsBuildSkipEagerWildCardEvaluationRegexes", ".*");
+
+            EngineFileUtilities.CaptureLazyWildcardRegexes();
+
+            Project project = Should.NotThrow(() => new Project(projectFiles.ProjectFile));
+
+            project.GetConcatenatedItemsOfType("None").ShouldContain("csc.*");
         }
 
         [Theory]

--- a/src/Build/Evaluation/LazyItemEvaluator.cs
+++ b/src/Build/Evaluation/LazyItemEvaluator.cs
@@ -436,14 +436,11 @@ namespace Microsoft.Build.Evaluation
 
             private static void ProcessNonWildCardItemUpdates(Dictionary<string, UpdateOperation> itemsWithNoWildcards, OrderedItemDataCollection.Builder items)
             {
-#if DEBUG
-                ErrorUtilities.VerifyThrow(itemsWithNoWildcards.All(fragment => !MSBuildConstants.CharactersForExpansion.Any(fragment.Key.Contains)), $"{nameof(itemsWithNoWildcards)} should not contain any text fragments with wildcards.");
-#endif
                 if (itemsWithNoWildcards.Count > 0)
                 {
                     for (int i = 0; i < items.Count; i++)
                     {
-                        string fullPath = FileUtilities.GetFullPath(items[i].Item.EvaluatedIncludeEscaped, items[i].Item.ProjectDirectory);
+                        string fullPath = FileUtilities.NormalizePathForComparisonNoThrow(items[i].Item.EvaluatedIncludeEscaped, items[i].Item.ProjectDirectory);
                         if (itemsWithNoWildcards.TryGetValue(fullPath, out UpdateOperation op))
                         {
                             items[i] = op.UpdateItem(items[i]);

--- a/src/Build/Evaluation/LazyItemEvaluator.cs
+++ b/src/Build/Evaluation/LazyItemEvaluator.cs
@@ -440,7 +440,7 @@ namespace Microsoft.Build.Evaluation
                 {
                     for (int i = 0; i < items.Count; i++)
                     {
-                        string fullPath = FileUtilities.NormalizePathForComparisonNoThrow(items[i].Item.EvaluatedIncludeEscaped, items[i].Item.ProjectDirectory);
+                        string fullPath = FileUtilities.NormalizePathForComparisonNoThrow(items[i].Item.EvaluatedInclude, items[i].Item.ProjectDirectory);
                         if (itemsWithNoWildcards.TryGetValue(fullPath, out UpdateOperation op))
                         {
                             items[i] = op.UpdateItem(items[i]);

